### DISCUSSION TITLE
Add more useful functions to the fields plugin

### DISCUSSION THREAD

**Discussion by 7sharp9** [MoiraeSoftware/Myriad#28] [Ideas]

**Comment by TysonMN** [comment]
For convenience in this issue, here is OCaml's [ppx_fields_conv](https://github.com/janestreet/ppx_fields_conv) plugin.

**Comment by TysonMN** [comment]
I would like the `get` functions named as such.  Though, I suppose both could exist.
```
let getOne r = r.One
```

I would like individual (copy-and-update) `set` functions.  Unlike `ppx_fields_conv`, I don't think it is a good idea to generate a function that mutates a field.  That way, there is more friction when deciding to go with a mutable approach.  However, if mutating functions are going to be generated, then I would prefer that they get a worse name than the copy-and-update function.  So the copy-and-update function is called `setOne` and the mutating function is called `mutateOne`.  I think this name matches well with F#'s `mutable` keyword.
```
let setOne v r = { r with One = v }
```

I would like individual `map` functions.
```
let mapOne f = { r with One = r.One |> f }
```

I recently decided that I can no longer live without these functions, so I started manually implementing them in field-specific modules.  That way, they are easy to copy-paste and find-replace when I add a new field.  It also makes reading the implementation of `map` a bit easier because then `map` is defined via `get` and `get` instead of having `mapOne` defined in terms of `getOne` and `setOne`.
```
let map get set f a =
  a |> get |> f |> Func.flip set a

type MyRecord =
  { MyField1: int
    MyField2: int }

module MyRecord =
  module MyField1 =
    let get m = m.MyField1
    let set v m = { m with MyField1 = v }
    let map f = f |> map get set
  module MyField2 =
    let get m = m.MyField2
    let set v m = { m with MyField2 = v }
    let map f = f |> map get set
```

Then calling code can open `MyRecord` and write `r |> MyField1.map increment`.

I don't have a strong preference for field-specific modules when generating the code though.  Also, a name like `mapOne` is more idiomatic (than `One.map`) in F# because of examples like `Result.mapError`.

I also think it would be interesting to consider defining `set` via `create` since this would allow a type parameter of a generic record to change because of a call to `set` or `map`.  However, this gets complicated fast.  For example, consider
```
type RoseTreeLike<'a, 'b> =
  { A: 'a
    B: 'b
    Children: RoseTreeLike<'a, 'b> list }
```

No individual field can change its type in a mapping, but `Children` and either `A` or `B` can vary their type (somewhat) at the same time.

I looked through the functions generated by `ppx_fields_conv`.  I don't see the value in them (yet!).

I currently see the value in individual `get`, `set`, and `map` functions though, and I would prefer to not have to keep doing copy-paste and find-replace.  I am interested in improving the fields plugin to make this happen.

**Comment by 7sharp9** [comment]
Im happy to take a PR adding anything that's useful as long as there is an example of usage or some documentation added.
